torch_xla/debug/metrics_test_wrapper.py

```diff
@@ -0,0 +1,84 @@
+"""Wrap PyTorch XLA jobs with extra logic around XLA metrics.
+
+It can be used to save a historical record of metrics and/or check metrics
+against a golden set of metrics to search for performance regressions.
+
+Example usage:
+python metrics_test_wrapper.py --metrics_output_filename="metrics.txt" \
+    -- python test/test_train_mnist.py --num_epochs=1
+"""
+import argparse
+import os
+import subprocess
+import sys
+
+import torch_xla.utils.gcsfs as gcsfs
+
+
+_CLOUD_STORAGE_PREFIX = 'gs://'
+_TMP_FILENAME = '/tmp/pytorch_tpu_metrics_tmp'
+_XLA_METRICS_FILE = 'XLA_METRICS_FILE'
+
+
+def _run_subprocess(cmd):
+  # Use _XLA_METRICS_FILE to pass the metrics report from the subprocess
+  # to this process.
+  os.environ[_XLA_METRICS_FILE] = _TMP_FILENAME
+  sp_return_code = subprocess.call(cmd)
+  with open(_TMP_FILENAME, 'r') as tmp_metrics:
+    metrics = tmp_metrics.read()
+  os.remove(_TMP_FILENAME)
+  del os.environ[_XLA_METRICS_FILE]
+
+  # Use only the metrics from the last step.
+  metrics = metrics[metrics.rfind('[MetricsData;'):]
+  return metrics, sp_return_code
+
+
+def _write_to_disk(output_string, output_filename,
+                   service_account_filename=None):
+  if not output_string:
+    return
+  try:
+    if output_filename.find(_CLOUD_STORAGE_PREFIX) == 0:
+      gcsfs.write(output_filename, output_string,
+                  service_account_filename=service_account_filename)
+    else:
+      with open(output_filename, 'w') as outfile:
+        outfile.write(output_string)
+  except Exception as e:
+    print('Failed to write metrics to file: {}'.format(e))
+
+
+if __name__ == '__main__':
+  parser = argparse.ArgumentParser(
+      description='PyTorch on TPU distrubuted training',
+      epilog=('Usage example: metrics_test_wrapper.py '
+          '--golden_metrics_filename="gcs://bucket/file" '
+          '--metrics_output_filename="gcs://bucket/file" '
+          '--service_account_filename="path/to/file" -- python train.py'))
+  parser.add_argument('--golden_metrics_filename', type=str, default=None)
+  parser.add_argument('--metrics_output_filename', type=str, default=None)
+  parser.add_argument('--service_account_filename', type=str, default=None)
+  parser.add_argument(
+      'positional',
+      nargs='+',
+      type=str,
+      help='The python command to run.')
+  FLAGS = parser.parse_args()
+  if not (FLAGS.golden_metrics_filename or FLAGS.metrics_output_filename):
```

Comment:
How things are going to work?
Will we have an outer wrapper script, or this is the `runner`?
If the latter, it should essentially take almost no argument besides the test command line, and the `test name`.
We can have a root GCS folder where all the per-test config is stored, which we append the test name, and inside there we expect to find certain file/folder structure (I know, GCS does not really have folders 😄 ).

Reply:
For the Google use case, we'll change our daily tests to call this wrapper and set these flags in that test config. I think it will be easier to debug issues if we have a code history of the config rather than a config file in GCS

Reply:
That is not a Google case.
That is a kind of config that usually is also global, and you do not change every API call.

Reply:
I'm saying that I'm imagining there will be an outer wrapper script. We will modify our current outer wrapper script to call `docker run python metrics_test_wrapper --metrics_output_filename="metrics.txt" --golden_metrics_filename="goldent.txt" --python test/test_train_mnist.py --num_epochs=1` instead of what it used to call: `docker run python test/test_train_mnist.py --num_epochs=1`. The outer script will take some action depending on return code

Reply:
This needs more configuration to work, and we will end up having to have another "thick" wrapper script.
Better this be The Script, modulo eventually a trivial shell script that sets up some environment and run this.

For every test, beside the test/model data, this script will need some input configuration.
So what I am saying is lets have a root folder, say `gs://pytorch-tpu-test/` and lets call this script as:
```
test_runner.py --root=gs://pytorch-tpu-test/ --test_name=resnet50_blah -- CMD ARGS...
```

Then the test_runner will find, inside the `gs://pytorch-tpu-test/resnet50_blah/` folder all the info it needs.
For example, there will be a `gs://pytorch-tpu-test/resnet50_blah/golden/` folder to store the goldens.
There might also, for example, a `gs://pytorch-tpu-test/resnet50_blah/config.json` to store test specific config (tolerances, notify emails, ...).
You get the idea.